src/celeritas/optical/detail/PreGenAction.hh

```diff
@@ -0,0 +1,111 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/detail/PreGenAction.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+
+#include "corecel/Macros.hh"
+#include "corecel/data/Collection.hh"
+#include "celeritas/global/ActionInterface.hh"
+#include "celeritas/optical/OpticalDistributionData.hh"
+
+namespace celeritas
+{
+class CerenkovParams;
+class OpticalPropertyParams;
+class ScintillationParams;
+
+namespace detail
+{
+struct OpticalGenStorage;
+//---------------------------------------------------------------------------//
+/*!
+ * Generate optical distribution data.
+ */
+class PreGenAction final : public ExplicitCoreActionInterface
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SPConstCerenkov = std::shared_ptr<CerenkovParams const>;
+    using SPConstProperties = std::shared_ptr<OpticalPropertyParams const>;
+    using SPConstScintillation = std::shared_ptr<ScintillationParams const>;
+    using SPGenStorage = std::shared_ptr<detail::OpticalGenStorage>;
+    //!@}
+
+    //! Check if the distribution data is valid
+    struct IsInvalid
+    {
+        CELER_FUNCTION bool operator()(OpticalDistributionData data) const
+        {
+            return !data;
+        }
+    };
+
+  public:
+    // Construct with action ID, optical properties, and storage
+    PreGenAction(ActionId id,
+                 SPConstProperties properties,
+                 SPConstCerenkov cerenkov,
+                 SPConstScintillation scintillation,
+                 SPGenStorage storage);
+
+    // Launch kernel with host data
+    void execute(CoreParams const&, CoreStateHost&) const final;
+
+    // Launch kernel with device data
+    void execute(CoreParams const&, CoreStateDevice&) const final;
+
+    //! ID of the model
+    ActionId action_id() const final { return id_; }
+
+    //! Short name for the action
+    std::string label() const final { return "optical-pre-generator-post"; }
```

Comment:
What does the trailing "-post" stand for (just for my curiosity)?

Reply:
The `pre-` is that it generates the "before an optical photon"; the `-post` is it's doing the calculation after the step (post-step).